lib/mix/tasks/compile.make.ex

```diff
@@ -99,69 +95,55 @@ defmodule Mix.Tasks.Compile.ElixirMake do
-  type in the following command (make sure that the path and version number
-  are correct):
-
-      cmd /K "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64
-
-  This should open up a command prompt with the necessary environment variables
-  set, and from which you will be able to run the "mix compile", "mix deps.compile",
-  and "mix test" commands.
-
-  Another option is to install the Linux compatiblity tools from [MSYS2](https://www.msys2.org/).
-
-  After installation start the msys64 bit terminal from the start menu and install the
-  C/C++ compiler toolchain. E.g.:
-
-    pacman -S --noconfirm pacman-mirrors pkg-config
-    pacman -S --noconfirm --needed base-devel autoconf automake make libtool git \
-      mingw-w64-x86_64-toolchain mingw-w64-x86_64-openssl mingw-w64-x86_64-libtool
-
-  This will give you a compilation suite nearly compatible with Unix' standard tools.
-  """
+  use Mix.Task
 
   @return if Version.match?(System.version(), "~> 1.9"), do: {:ok, []}, else: :ok
 
-  @spec run(OptionParser.argv()) :: :ok | no_return
   def run(args) do
     config = Mix.Project.config()
-    Mix.shell().print_app()
-    priv? = File.dir?("priv")
-    Mix.Project.ensure_structure()
-    build(config, args)
-
-    # IF there was no priv before and now there is one, we assume
-    # the user wants to copy it. If priv already existed and was
-    # written to it, then it won't be copied if build_embedded is
-    # set to true.
-    if not priv? and File.dir?("priv") do
-      Mix.Project.build_structure()
+    app = config[:app]
+    version = config[:version]
+    force_build = pre_release?(version) or Keyword.get(config, :make_force_build, false)
+    precompiler = config[:make_precompiler]
+
+    if force_build == true or precompiler == nil do
+      Mix.Tasks.ElixirMake.Precompile.build_native(args)
+    else
+      context = precompiler_context(args, precompiler)
+      nif_filename = config[:make_nif_filename] || "#{app}"
+      priv_dir = ElixirMake.Artefact.app_priv(app)
+
+      load_path =
+        case :os.type() do
+          {:win32, _} -> Path.join([priv_dir, "#{nif_filename}.dll"])
+          _ -> Path.join([priv_dir, "#{nif_filename}.so"])
+        end
+
+      with {:skip_if_exists, false} <- {:skip_if_exists, File.exists?(load_path)},
+           {:error, precomp_error} <-
+             Mix.Tasks.ElixirMake.Precompile.download_or_reuse_nif_file(context) do
+        message = """
+        Error while downloading precompiled NIF: #{precomp_error}.
+        You can force the project to build from scratch with:
+            mix elixir_make.precompile
+        """
+
+        Mix.raise(message)
+        :error
+      else
+        _ -> @return
+      end
     end
+  end
+
+  defp precompiler_context(args, module) do
+    module = Module.concat([Mix.Tasks.ElixirMake, module])
 
-    @return
+    if Code.ensure_loaded?(module) do
+      Kernel.apply(module, :precompiler_context, [args])
```

Comment:
You declare this callback as optional, but here we always invoke it. Should we not make it optional?

Reply:
Actually we never directly use the context, so I have removed it. :)